src/Optional.js

```diff
@@ -36,7 +37,13 @@ class Optional {
 export const optional = curry((preview, set) => new Optional(preview, set))
 
 // optionalProp : String → Optional s a
-export const optionalProp = (key) => optional(get(key), set(key))
-
-// optionalIndex : Number → Optional s a
-export const optionalIx = (index) => optional(get(index), set(index))
+export const optionalProp = (key) => optional(
+  (obj)      => key in obj ? obj[key] : notFound,
+  (val, obj) => key in obj ? { ...obj, [key]: val } : obj
+)
+
+// optionalIx : Number → Optional s a
+export const optionalIx = (index) => optional(
+  (obj)      => index in obj ? obj[index] : notFound,
+  (val, obj) => index in obj ? { ...obj, [index]: val } : obj
```

Comment:
This is a very haskelly formatting... we have prettier for this 😝  please run it!